onnxruntime/core/providers/qnn/builder/qnn_backend_manager.cc

```diff
@@ -201,7 +269,240 @@ Status QnnBackendManager::ReleaseContext() {
 }
 
-Status QnnBackendManager::SetupBackend(const logging::Logger& logger) {
+Status QnnBackendManager::DumpQnnContext(const onnxruntime::PathString& context_cache_pathstring,
+                                         const std::string& model_name,
+                                         const std::string& graph_name) {
+  if (nullptr == qnn_interface_.contextGetBinarySize ||
+      nullptr == qnn_interface_.contextGetBinary) {
+    LOGS(*logger_, ERROR) << "Failed to get valid function pointer.";
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Failed to get valid function pointer.");
+  }
+
+  uint64_t required_buffer_size(0);
+  Qnn_ErrorHandle_t rt = qnn_interface_.contextGetBinarySize(context_, &required_buffer_size);
+  if (QNN_CONTEXT_NO_ERROR != rt) {
+    LOGS(*logger_, ERROR) << "Failed to get QNN context binary size. Error code: " << rt;
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Failed to get QNN context binary size.");
+  }
+
+  std::unique_ptr<unsigned char[]> context_buffer = std::make_unique<unsigned char[]>(required_buffer_size);
+  if (nullptr == context_buffer) {
+    LOGS(*logger_, ERROR) << "Failed to allocate buffer for context cache.";
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Failed to allocate buffer for context cache.");
+  }
+
+  uint64_t written_buffer_size(0);
+  rt = qnn_interface_.contextGetBinary(context_,
+                                       reinterpret_cast<void*>(context_buffer.get()),
+                                       required_buffer_size,
+                                       &written_buffer_size);
+  if (QNN_CONTEXT_NO_ERROR != rt) {
+    LOGS(*logger_, ERROR) << "Failed to get context binary.";
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Failed to get context binary.");
+  }
+
+  if (required_buffer_size < written_buffer_size) {
+    LOGS(*logger_, ERROR) << "Context written buffer size: " << written_buffer_size
+                          << " exceeds allocated buffer size: " << required_buffer_size;
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Context written buffer exceeds allocated buffer size.");
+  }
+
+  std::ofstream of_stream(context_cache_pathstring.c_str(), std::ofstream::binary);
+  if (!of_stream) {
+    LOGS(*logger_, ERROR) << "Failed to open cached context file.";
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Failed to open context cache file.");
+  }
+
+  // Reseve 255 bytes to add metadata into context binary file
+  auto length_required = strlen(PROVIDER_KEY) + strlen(MODEL_NAME_KEY) + strlen(GRAPH_NAME_KEY) +
+                         strlen(QNN_PROVIDER) + model_name.length() + graph_name.length() + 7;
+  ORT_ENFORCE(length_required <= RESERVED_LENGTH, "Model name too long, exceeds context binary metadata length limit.");
+  char context_buffer_meta[RESERVED_LENGTH];
+  snprintf(context_buffer_meta, RESERVED_LENGTH, "{%s:%s|%s:%s|%s:%s}",
+           PROVIDER_KEY, QNN_PROVIDER, MODEL_NAME_KEY, model_name.c_str(), GRAPH_NAME_KEY, graph_name.c_str());
```

Comment:
I don't think we need to serialize keys. Since we are defining the format of this header, we can enforce that information is stored in an explicit/fixed order. This will make the header smaller (leaves more room for model name) and will make parsing simpler. This is what is done for most headers in various file formats (png, exe, elf, etc.)

For example:
```

    magic bytes           total header size      model name len (X)       model name          graph name len (Y)         graph name

| ------ 8 bytes ------| ------ 2 bytes ----- | ------ 2 bytes ----- | ------ X bytes ---- | ------- 2 bytes ------- | ----- Y bytes ----|
'O''R''T''Q''N''N''E''P'        0x00FF                 0x000F          'r''e''s''n''e''t'            0x0008                   'Q'..

length_required = 8 + 2 + 2 + model_name.length + 2 + graph_name.length()

```

- If the magic bytes don't match exactly, then we can assume this is not generated by ORT QNN EP.
- If the magic byte do match, we read the next 2 bytes and allocate a buffer of that size.
- Then we read the entire header into the buffer.

Also, a maximum header size of 255 bytes may be too small for longer model names. It would be inconvenient for the user to have to rename the model when we could just bump up the maximum (e.g., 1024 bytes). Because the above header specifies the sizes of everything (including the total header size), its size can be flexible (e.g., small model name has a smaller header).

Reply:
I should note that the above format requires us to be more careful with byte order (endianness) of the header. We can define that the header is always stored in little-endian order. We have utils for reading/writing in https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/framework/endian_utils.h

Reply:
updated.

---
In reply to: [1198174690](https://github.com/microsoft/onnxruntime/pull/15815#discussion_r1198174690) [](ancestors = 1198174690)

Reply:
updated

---
In reply to: [1198122293](https://github.com/microsoft/onnxruntime/pull/15815#discussion_r1198122293) [](ancestors = 1198122293)